k2/csrc/fsa_algo.cc

```diff
@@ -275,6 +275,111 @@ bool Connect(const Fsa &a, Fsa *b, std::vector<int32_t> *arc_map /*=nullptr*/) {
   return is_acyclic;
 }
 
+void RmEpsilonsPrunedMax(const WfsaWithFbWeights &a, float beam, Fsa *b,
+                         std::vector<std::vector<int32_t>> *arc_derivs) {
+  CHECK_EQ(a.weight_type, kMaxWeight);
+  CHECK_GT(beam, 0);
+  CHECK_NOTNULL(b);
+  CHECK_NOTNULL(arc_derivs);
+  b->arc_indexes.clear();
+  b->arcs.clear();
+  arc_derivs->clear();
+
+  const auto &fsa = a.fsa;
+  if (IsEmpty(fsa)) return;
+  int32_t num_states_a = fsa.NumStates();
+  int32_t final_state = fsa.FinalState();
+  const auto &arcs_a = fsa.arcs;
+  const float *arc_weights_a = a.arc_weights;
+
+  // identify all states that should be kept
+  std::vector<char> non_eps_in(num_states_a, 0);
+  non_eps_in[0] = 1;
+  using WeightsPair = std::pair<double, std::vector<int32_t>>;
+  // (label, dest_state) -> `sum` of weights along all paths from current state
+  // to `dest_state` with label == `label`(or plus numbers of epsilon),
+  // `vector<int32_t>` in `WeightsPair` records all arc-indexes in `a` that
+  // contributes the current arc in `b`.
+  using ArcMap =
+      std::unordered_map<std::pair<int32_t, int32_t>, WeightsPair, PairHash>;
+  std::vector<ArcMap> arcs_b(num_states_a);
+  for (int32_t i = static_cast<int32_t>(arcs_a.size()) - 1; i >= 0; --i) {
+    const auto &arc = arcs_a[i];
+    const auto src_state = arc.src_state;
+    const auto dest_state = arc.dest_state;
+    const auto label = arc.label;
+    DCHECK_GE(dest_state, src_state);
+
+    double arc_weight = arc_weights_a[i];
+    if (label != kEpsilon) {
```

Comment:
This algorithm would take quadratic time in the length of epsilon paths.  Better to use a traceback-based algorithm, like what I sketched in the header.

Reply:
if there is path `a->b->c->d` and `e->b->c->d` with all labels are epsilon, the `traceback-based` algorithm will iterate over `b->c->d` twice, right? And the current implementation will only iterate over the path one time (Of course with a cost of copying non-epsilon output arcs for those states that has epsilon arcs in).

Fo example,  if there is copy arc in the current implementation (suppose a->b with label epsilon (no label), and b->c with label `l`, then there is a  copy arc  `a->c` with label `l`), then there's also a trace path from a->b->c (and may multiple such iterations as there maybe `e->a` with epsilon and `f->a` with epsilon) in traceback-based algorithm, so seems the current implenmentation will  take less time?

Did I express clearly? Correct me If I'm wrong.

Reply:
The problem is that your algorithm can in general be quadratic in the size of the output FSA, because for a sequence of epsilons, you have to append those vector<int> things.  Mine is linear in the size of the output FSA, assuming the input is connected.

Reply:
The problem is that your algorithm can in general be quadratic in the size of the output FSA, because for a sequence of epsilons, you have to append those vector<int> things.  Mine is linear in the size of the output FSA, assuming the input is connected.

Reply:
.. each time you trace e->a or f->a or the like, you are taking time O(length of that sequence a->b->c->d) as you are manipulating those vector<int> things.

Reply:
Oh, ok, will revert back to the traceback-based algorithm.